lib/grape/validations/types/custom_type_collection_coercer.rb

```diff
@@ -0,0 +1,71 @@
+module Grape
+  module Validations
+    module Types
+      # Instances of this class may be passed to
+      # +Virtus::Attribute.build+ as the +:coercer+
+      # option, to handle collections of types that
+      # provide their own parsing (and optionally,
+      # type-checking) functionality.
+      #
+      # See {CustomTypeCoercer} for details on types
+      # that will be supported by this by this coercer.
+      # This coercer works in the same way as +CustomTypeCoercer+
+      # except that it expects to receive an array of strings to
+      # coerce and will return an array (or optionally, a set)
+      # of coerced values.
+      #
+      # +CustomTypeCoercer+ is already capable of providing type
+      # checking for arrays where an independent coercion method
+      # is supplied. As such, +CustomTypeCollectionCoercer+ does
+      # not allow for such a method to be supplied independently
+      # of the type.
+      class CustomTypeCollectionCoercer < CustomTypeCoercer
+        # A new coercer for collections of the given type.
+        #
+        # @param type [Class,#parse]
+        #   type to which items in the array should be coerced.
+        #   Must implement a +parse+ method which accepts a string,
+        #   and for the purposes of type-checking it may either be
+        #   a class, or it may implement a +coerced?+, +parsed?+ or
+        #   +call+ method (in that order of precedence) which
+        #   accepts a single argument and returns true if the given
+        #   array item has been coerced correctly.
+        # @param set [Boolean]
+        #   when true, a +Set+ will be returned by {#call} instead
+        #   of an +Array+ and duplicate items will be discarded.
+        def initialize(type, set = false)
+          super(type)
+          @set = set
+        end
+
+        # This method is called from somewhere within
+        # +Virtus::Attribute::coerce+ in order to coerce
+        # the given value.
+        #
+        # @param value [Array<String>] an array of values to be coerced
+        # @return [Array,Set] the coerced result. May be an +Array+ or a
+        #   +Set+ depending on the setting given to the constructor
+        def call(value)
+          coerced = value.map { |item| super(item) }
```

Comment:
I'm aware this will fail if `value` is a string but I wasn't sure of the best thing to do about it. Maybe `Array.wrap`?

Reply:
To allow for a single item to be passed in as such (without the array wrapping/declaration) is a convenience, but could be improper. I suppose it depends on how strict Grape wants to be.

Reply:
I would deal with that scenario later with a proper spec.